pkg/kv/kvserver/rangefeed/buffered_sender.go

```diff
@@ -51,54 +56,130 @@ type BufferedSender struct {
 	// thread safety.
 	sender ServerStreamSender
 
-	// metrics is used to record rangefeed metrics for the node.
-	metrics RangefeedMetricsRecorder
+	queueMu struct {
+		syncutil.Mutex
+		stopped bool
+		buffer  *eventQueue
+	}
+
+	// Unblocking channel to notify the BufferedSender.run goroutine that there
+	// are events to send.
+	notifyDataC chan struct{}
 }
 
-func NewBufferedSender(
-	sender ServerStreamSender, metrics RangefeedMetricsRecorder,
-) *BufferedSender {
-	return &BufferedSender{
-		sender:  sender,
-		metrics: metrics,
+func NewBufferedSender(sender ServerStreamSender) *BufferedSender {
+	bs := &BufferedSender{
+		sender: sender,
 	}
+	bs.queueMu.buffer = newEventQueue()
+	bs.notifyDataC = make(chan struct{}, 1)
+	return bs
 }
 
 // SendBuffered buffers the event before sending them to the underlying
 // ServerStreamSender.
-func (bs *BufferedSender) SendBuffered(
-	event *kvpb.MuxRangeFeedEvent, alloc *SharedBudgetAllocation,
-) error {
-	panic("unimplemented: buffered sender for rangefeed #126560")
+//
+// alloc.Release is nil-safe. SendBuffered will take the ownership of the alloc
+// and release it if the return error is non-nil. Note that it is safe to send
+// error events without being blocked for too long.
+func (bs *BufferedSender) send(ev *kvpb.MuxRangeFeedEvent, alloc *SharedBudgetAllocation) error {
+	// Make sure this is from catch up scan.
+	if ev.Error != nil && alloc != nil {
+		return bs.sendUnbuffered(ev)
+	}
+	return bs.sendBuffered(ev, alloc)
 }
 
-// SendUnbuffered bypasses the buffer and sends the event to the underlying
-// ServerStreamSender directly. Note that this can cause event re-ordering.
-// Caller is responsible for ensuring that events are sent in order.
-func (bs *BufferedSender) SendUnbuffered(
-	event *kvpb.MuxRangeFeedEvent, alloc *SharedBudgetAllocation,
+func (bs *BufferedSender) sendBuffered(
+	ev *kvpb.MuxRangeFeedEvent, alloc *SharedBudgetAllocation,
 ) error {
-	panic("unimplemented: buffered sender for rangefeed #126560")
+	bs.queueMu.Lock()
+	defer bs.queueMu.Unlock()
+	if bs.queueMu.stopped {
+		log.Errorf(context.Background(), "stream sender is stopped")
```

Comment:
The caller can log this error if they care to.

Reply:
Removed.